### PR TITLE
Issue 2044 - show skipping reason in verbose mode

### LINKED ISSUE #2044: add option to see skip reason and verbose
```$ pip list
beautifulsoup4 (4.4.1)
chardet (2.3.0)
defer (1.0.6)
html5lib (0.999)
lxml (3.5.0)
pip (8.1.1)
psutil (3.4.2)
py (1.4.31)
pygobject (3.20.0)
pysqlite (2.7.0)
pytest (2.8.7)
python-apt (1.1.0b1)
python-debian (0.1.27)
pyxdg (0.25)
setuptools (20.7.0)
six (1.10.0)
vboxapi (1.0)
wheel (0.29.0)

platform linux2 -- Python 2.7.12, pytest-2.8.7, py-1.4.31, pluggy-0.3.1 -- /usr/bin/python


$ py.test -v testscript.py
============================= test session starts ==============================
platform linux2 -- Python 2.7.12, pytest-2.8.7, py-1.4.31, pluggy-0.3.1 -- /usr/bin/python
cachedir: .cache
rootdir: /home/rjl/pytest, inifile:
collected 8 items

testscript.py::TestAclass::test_1 PASSED
testscript.py::TestAclass::test_2 PASSED
testscript.py::TestAclass::test_3 PASSED
testscript.py::TestAclass::test_4 PASSED
testscript.py::TestBclass::test_21 SKIPPED
testscript.py::TestBclass::test_22 SKIPPED
testscript.py::TestBclass::test_23 SKIPPED
testscript.py::TestBclass::test_24 SKIPPED

===================== 4 passed, 4 skipped in 0.03 seconds ======================

$ py.test -rv testscript.py
============================= test session starts ==============================
platform linux2 -- Python 2.7.12, pytest-2.8.7, py-1.4.31, pluggy-0.3.1
rootdir: /home/rjl/pytest, inifile:
collected 8 items

testscript.py ....ssss

===================== 4 passed, 4 skipped in 0.01 seconds ======================

$ py.test -rs testscript.py
============================= test session starts ==============================
platform linux2 -- Python 2.7.12, pytest-2.8.7, py-1.4.31, pluggy-0.3.1
rootdir: /home/rjl/pytest, inifile:
collected 8 items

testscript.py ....ssss
=========================== short test summary info ============================
SKIP [4] /home/rjl/pytest/conftest.py:8: test needs -A option to run

===================== 4 passed, 4 skipped in 0.01 seconds ======================

```

I want the output from the first one showing 'PASSED' or 'SKIPPED' but I also want to see the reason a specific test was skipped.

Something like:

```
$ py.test -v testscript.py
============================= test session starts ==============================
platform linux2 -- Python 2.7.12, pytest-2.8.7, py-1.4.31, pluggy-0.3.1 -- /usr/bin/python
cachedir: .cache
rootdir: /home/rjl/pytest, inifile:
collected 8 items

testscript.py::TestAclass::test_1 PASSED
testscript.py::TestAclass::test_2 PASSED
testscript.py::TestAclass::test_3 PASSED
testscript.py::TestAclass::test_4 PASSED
testscript.py::TestBclass::test_21 SKIPPED (needs -A option to run)
testscript.py::TestBclass::test_22 SKIPPED (needs -A option to run)
testscript.py::TestBclass::test_23 SKIPPED (needs -A option to run)
testscript.py::TestBclass::test_24 SKIPPED (needs -A option to run)
```

### PR DESCRIPTION
**Pull request by CarycaKatarzyna**
I've implemented functionality proposed in #2044. 

Fixes #2044.

### CODE REVIEW

**Comment by CarycaKatarzyna** [comment]
Would anyone like to add review? Please don't hesitate to write if you dislike the proposed functionality.

File: testing/test_terminal.py [hunk 1]
```diff
@@ -2177,3 +2200,20 @@ def test_foo():
                 ]
             )
         )
+
+
+def test_raw_skip_reason():
```
Comment (nicoddemus) [inline testing/test_terminal.py:2205]: 👍

File: testing/test_terminal.py [hunk 2]
```diff
@@ -2177,3 +2200,20 @@ def test_foo():
...
+
+    class X:
+        pass
+
+    ev = X()
+    ev.skipped = True
+    ev.longrepr = longrepr
+
+    reason = _raw_skip_reason(ev)
+    assert reason == message[9:]
```
Comment (nicoddemus) [inline testing/test_terminal.py:2219]:
```suggestion
    assert reason == "Just so"
```

File: src/_pytest/terminal.py [hunk 3]
```diff
@@ -1230,3 +1233,12 @@ def format_session_duration(seconds: float) -> str:
...
         dt = datetime.timedelta(seconds=int(seconds))
         return "{:.2f}s ({})".format(seconds, dt)
+
+
+def _raw_skip_reason(rep):
+    assert rep.skipped
+    assert len(rep.longrepr) == 3
+    _, _, reason = rep.longrepr
+    if reason.startswith("Skipped: "):
+        reason = reason[9:]
```
Comment (nicoddemus) [inline src/_pytest/terminal.py:1243]:
```suggestion
        reason = reason[len("Skipped: "):]
```

Just for clarity. :+1:

(Probably worthwhile moving `"Skipped: "` to a variable and reuse in both places).

File: src/_pytest/terminal.py [hunk 4]
```diff
@@ -483,6 +483,9 @@ def pytest_runtest_logreport(self, report: TestReport) -> None:
             line = self._locationline(rep.nodeid, *rep.location)
             if not running_xdist:
                 self.write_ensure_prefix(line, word, **markup)
+                if rep.skipped and not was_xfail:
+                    reason = _raw_skip_reason(rep)
```
[the remark this answers is not in the document]
Comment (CarycaKatarzyna) [reply src/_pytest/terminal.py]: Good idea! I'm going to add it soon.

**Comment by gnikonorov** [comment]
In response to @nicoddemus' comment above I agree we should add this for `xfail`, but perhaps we should raise a follow up issue for it so that this can go in. The changes aren't dependent on one another.

**Comment by nicoddemus** [comment]
> In response to @nicoddemus' comment above I agree we should add this for xfail, but perhaps we should raise a follow up issue for it so that this can go in. The changes aren't dependent on one another.

They are not, and while I usually agree it is better to make follow ups instead of bloating pull requests, I feel that in this case not adding the same support for `xfail` feels like a missing/incomplete feature, so I would prefer to add it here if @CarycaKatarzyna doesn't mind.

**Comment by gnikonorov** [comment]
> > In response to @nicoddemus' comment above I agree we should add this for xfail, but perhaps we should raise a follow up issue for it so that this can go in. The changes aren't dependent on one another.
> 
> They are not, and while I usually agree it is better to make follow ups instead of bloating pull requests, I feel that in this case not adding the same support for `xfail` feels like a missing/incomplete feature, so I would prefer to add it here if @CarycaKatarzyna doesn't mind.

Sorry @nicoddemus - thought they were independent

**Comment by CarycaKatarzyna** [comment]
I've followed your comments, but found some inconsistency during testing. Please see an example:
```bash
# test_1.py
import pytest
def test_function():
    pytest.xfail()

# test_2.py
import pytest 
@pytest.mark.skip
def test_function():
    pass

# test_3.py
import pytest 
def test_function():
    pytest.skip()
```
The result of calling ``pytest -v``  is:
```
test_1.py::test_function XFAIL ()                                                                    [ 33%]
test_2.py::test_function SKIPPED (unconditional skip)                                                [ 66%]
test_3.py::test_function SKIPPED (<Skipped instance>)                                                [100%]
```
It means that reasons of skipping are different in all files. Shouldn't be the same or similar?

**Comment by bluetech** [comment]
I applied the changes mentioned above, rebased and also added support for XPASSED, so I think this should be good to go.

**Review by bluetech — APPROVED** [review]

File: testing/test_terminal.py [hunk 5]
```diff
@@ -2312,3 +2343,19 @@ def test_foo():
                 ]
             )
         )
+
+
+def test_raw_skip_reason() -> None:
```
Comment (nicoddemus) [inline testing/test_terminal.py:2348]: Perhaps add a similar test for the `wasxfail` case?

**Review by nicoddemus — APPROVED** [review]
LGTM, thanks @bluetech for picking it up!

File: testing/test_terminal.py (hunk 5 above)
Comment (bluetech) [reply testing/test_terminal.py]: Added, thanks.

**Pull request merged**